test/lock_free_queue_tests.cpp

```diff
@@ -0,0 +1,487 @@
+  for (auto i = 1; i <= 10; i++) {
+    ASSERT_TRUE(buffer.push(i)) << "Buffer should have space for element as size is 10";
+    ASSERT_EQ(i, buffer.size());
+  }
+  double obj1;
+  ASSERT_TRUE(buffer.get_latest(obj1));
+  ASSERT_EQ(10, obj1);
+  ASSERT_TRUE(buffer.empty()) << "Buffer should be empty";
+  ASSERT_EQ(0, buffer.size());
+  ASSERT_FALSE(buffer.get_latest(obj1));
+}
+
+TEST(LockFreeSPSCQueue, test_bounded_push)
+{
+  {
+    LockFreeSPSCQueue<double, 10> queue;
+    ASSERT_TRUE(queue.empty()) << "Buffer should be empty";
+    for (auto i = 1; i <= 25; i++) {
+      ASSERT_TRUE(queue.bounded_push(i)) << "Buffer should have space for element as size is 10";
+      ASSERT_EQ(std::min(i, 10), queue.size());
+      ASSERT_EQ(10, queue.capacity());
+    }
+    ASSERT_EQ(10, queue.size());
+    ASSERT_EQ(10, queue.capacity());
+
+    // when we start popping, the pop-ed elements should start from 16
+    for (auto i = 1u; i <= queue.capacity(); i++) {
+      double obj1;
+      ASSERT_TRUE(queue.pop(obj1));
+      ASSERT_EQ(i + 15, obj1);
+      ASSERT_EQ(queue.capacity() - i, queue.size());
+    }
+    double obj1;
+    ASSERT_FALSE(queue.pop(obj1));
+    ASSERT_TRUE(queue.empty()) << "Buffer should be empty";
+  }
+  {
+    LockFreeSPSCQueue<double> queue(10);
+    ASSERT_TRUE(queue.empty()) << "Buffer should be empty";
+    for (auto i = 1; i <= 25; i++) {
+      ASSERT_TRUE(queue.bounded_push(i)) << "Buffer should have space for element as size is 10";
+      ASSERT_EQ(std::min(i, 10), queue.size());
+      ASSERT_EQ(10, queue.capacity());
+    }
+    ASSERT_EQ(10, queue.size());
+    ASSERT_EQ(10, queue.capacity());
+
+    // when we start popping, the pop-ed elements should start from 16
+    for (auto i = 1u; i <= queue.capacity(); i++) {
+      double obj1;
+      ASSERT_TRUE(queue.pop(obj1));
+      ASSERT_EQ(i + 15, obj1);
+      ASSERT_EQ(queue.capacity() - i, queue.size());
+    }
+    double obj1;
+    ASSERT_FALSE(queue.pop(obj1));
+    ASSERT_TRUE(queue.empty()) << "Buffer should be empty";
+  }
+}
+
+TEST(LockFreeSPSCQueue, test_lockfree_queue_push)
+{
+  LockFreeSPSCQueue<int> queue(100);
+  int producer_count = 0;
+  std::atomic_int consumer_count(0);
+  std::atomic_bool done(false);
+
+  const int iterations = 1000000;
+
+  std::thread producer([&]() {
+    for (int i = 0; i < iterations; ++i) {
+      while (!queue.push(i)) {
+        std::this_thread::yield();
+      }
+      ++producer_count;
+    }
+  });
+
+  std::thread consumer([&]() {
+    int value;
+    while (!done) {
+      while (queue.pop(value)) {
+        ++consumer_count;
+      }
+    }
+    while (queue.pop(value)) {
+      ++consumer_count;
+    }
+  });
+
+  producer.join();
+  done = true;
+  consumer.join();
+
+  ASSERT_EQ(producer_count, consumer_count);
+  ASSERT_EQ(producer_count, iterations);
+  ASSERT_EQ(consumer_count, iterations);
+}
+
+// TEST(LockFreeSPSCQueue, test_lockfree_queue_bounded_push)
```

Comment:
?

Reply:
This is the test that needs to be removed for multithreaded bounded queue for SPSC Queue. I'll do the cleanup now

Reply:
Done